### ISSUE TITLE
Citator Experiments: Fine-grained classification

### ISSUE THREAD

**Issue by rachlllg** [freelawproject/courtlistener#5077]
All prior experiments will be conducted on (simpler) binary classification (overruled or not overruled), once Eyecite enhancements are in place, hopefully the excerpts will include all the details for the model to make more fine-grained classifications (followed, distinguished, partially overruled, fully overruled, rejected, mentioned, etc). 

This is a blanket ticket for experiments needed to achieve this. Ideally, we will use the same model as we selected for binary classification but enhance it by tuning the prompt and/or finetuning the model to achieve this.

**Comment by rachlllg** [comment]
Looked at this briefly to get the ball rolling

- Considered http://scdb.wustl.edu/documentation.php?var=caseDisposition, but this dataset is only the final opinion of the court (at the end of the opinion, the word that says the court's treatment for the current case). Our use case differs, as we need to identify of the court's treatment towards another case. This, however, might be a useful set for appellate cases, where the final treatment is the treatment towards the appealed/cited case.
- I have not looked at any of the shepard books, but how about the below treatment list?
   - Stop:
      - Overruled
      - Reversed
   - Caution:
      - Criticized
      - Questioned
      - Limited
   - Neutral:
      - Cited
      - Distinguished
      - Explained
- I propose for the initial experiments, we start with the same scotus dataset we've been working on, use gpt to assign above labels to the cited opinions for human to review/confirm. This includes 133 citing opinions that cites ~1K cited opinions, which is a good starting point, and we already have the "overruled" labels so it saves a bit extra work.

**Comment by mlissner** [comment]
The approach sounds fine. I think Rebecca would be the best person to ask about the categories though.

**Comment by rachlllg** [comment]
Updated per Rebecca's notes as a starting point:
- Stop:
   - Overruled
   - Reversed
- Caution:
   - Criticized
   - Questioned
   - Limited
   - Distinguished
- Neutral:
   - Cited
   - Explained
- Uncertain:
   - These are the ones where the prediction is not certain enough for us to present a classification to the user
   - We need to think about how to determine these, since the model confidence score from a generative model is not deterministic

The next step is to use a model to take a first pass at generating these labels for our dataset, review the predictions & tweak the prompts for our selected model to experiment for performance.

**Comment by rachlllg** [comment]
> We need to think about how to determine these, since the model confidence score from a generative model is not deterministic

We can consider using a model as evaluator, or using two models concurrently.
- Model as evaluator: use another model and a prompt crafted for evaluation to predict whether the model prediction is correct or not correct
- Two models concurrently: use two models and the same prompts to make separate predictions, if the predictions agree, then we infer the classification is "confident"
- Using quotation also requires another model as evaluator to evaluate whether the quoted sentence indeed implies the classification

Add'l thoughts:
Rebecca shared the 2016 paper "Yellow Flag Fever: Describing Negative Legal Precedent in Yellow Flag Fever: Describing Negative Legal Precedent in Citators". The author worked on FastCase Bad Law Bot, a few thoughts:
-  FastCase used an algorithm to "search for instances of negative sub-sequent case treatment using “explanatory phrases” from table 8 of the Bluebook." 
   - Is Bluebook still used today? How can I get my hands on table 8 of the Bluebook?
- FastCase used their algorithm to identify cases that require manual review and their staffs used KeyCite to assist in their review
   - Would this be an acceptable approach for us to consider when we are reviewing the model predictions? We are not using their data to train any models, since the predictions are made based on the quality of the prompt, not on any training data in itself.
- An additional way of identifying the treatment of a cited case, which I hadn't thought of before, is utilizing a related opinion (A) that discusses how the acting/citing opinion (B) treated the target/cited opinion (C), without having explicitly applied a treatment to the cited opinion (C). 
   - We should include this, as these are judge's interpretations, particularly when the treatment of B towards C is unclear or difficult to determine based on the body of opinion B itself
   - But the relationship of "B treats C" can be discussed by numerous related opinions, so we need to apply some algorithm to determine the true treatment (based on frequency, recency, court authority, etc)
- We should also utilize hierarchical labels, as we suggested above, from most negative to somewhat negative. Also tally the treatments so it's immediately apparent which ones should be cautioned against.
- I used gpt-4.1 to generate labels on the dataset, now I'm reviewing and updating them, I already discussed with Rebecca and she will sample some to review next week while I'm out.

[Yellow Flag Fever_ Describing Negative Legal Precedent in Citator.pdf](https://github.com/user-attachments/files/20236636/Yellow.Flag.Fever_.Describing.Negative.Legal.Precedent.in.Citator.pdf)

**Comment by rachlllg** [comment]
To generate labels for the fine-grained treatments, I crafted a [new prompt](https://github.com/freelawproject/ai-experiments/blob/FLP_Citator/experiments_515/utils/instructions.py) with definitions of the treatments and provided the prompt along with the citing opinion and cited opinions metadata to 3 models to generate labels.  The models are `gpt-4.1-2025-04-14`, `anthropic.claude-3-5-sonnet-20241022-v2:0`, and `us.anthropic.claude-sonnet-4-20250514-v1:0`.

The final labels are assigned based on below logic:
1. If in congress.gov dataset as overruled --> overruled
2. If Rebecca reviewed --> Use Rebecca assigned label
3. If all three models made the same prediction --> Use model prediction
4. If at two models made the same prediction --> Use model prediction
5. Otherwise, I manually reviewed and assigned labels based on my best judgment
6. I also selected a number of citing opinions for manual review based on my best judgment

Rebecca reviewed 105 instances while I reviewed 95 instances. The labels generated from steps 1 and 2 are considered golden labels (column "True_Labelled"). The labels generated from all subsequent steps require expert review and are used only as an estimate for this experiment.

Since GPT is the only model that allows users to enforce json output format, it is the only model that generated predictions for all our records and therefore the evaluation is run on GPT predictions only, below is the evaluation results.

- The model did decently well with "Overruled" (F1=0.85), with room for improvement to recall
- The model did quite well with "Mentioned (F1=0.97), this is not surprising as these are often clear languages where the citing opinion simply make a statement and cites numerous decisions that support the statement
- The model also did quite well with "Reversed" (F1= 0.88 with both high precision & recall), this is expected as SCOTUS opinions often make it very clear the case being reviewed and reversed in the body of the opinion
- The model did decently well with "Distinguished" (F1=0.80) where false predictions are confusion with "Limited", "Mentioned", and "Overruled". This is also something I personally find somewhat confusing so SME guidance would be needed to enhance the prompt.
- The model did poorly on "Criticized" (F1=0.48 with 21 supports) and "Limited" (F1=0.32 with 17 supports) despite a decent number of support, again, SME guidance would be needed to enhance the prompt.
- We do not have enough examples of "Questioned" to produce a meaningful metric, so more samples are needed.


![Image](https://github.com/user-attachments/assets/0033ca6d-d479-4076-9316-0baa61ce4821)
![Image](https://github.com/user-attachments/assets/e0b935ed-0109-46b3-9414-32bb575a50e0)

Below is the eval on the golden labels, however, this is not a very meaningful result as we do not have enough examples for each class to generate a meaningful eval metric, more SME-reviewed samples are needed.

![Image](https://github.com/user-attachments/assets/5a81f01a-eb21-43d6-8a25-305ed9a57c48)
![Image](https://github.com/user-attachments/assets/86b60d37-4ddd-4a17-8748-3c2a3a96e0a6)

# Conclusion:
The current approach is a bit "backwards" from the final product we want to produce. Currently, we analyze the cited opinions within a particular citing opinion (ie, we are assigning treatment to the opinions on the "authority" page), but the final product should be analyzing the treatment of a cited opinion amongst all subsequent citing opinions (ie, assigning treatment to the opinions on the "cited by" page). This distinction was not apparent to me previously and I'm not entirely sure why we chose this direction instead of the other one (perhaps because this one is easier to solve?) but I think we should move towards the final product direction. Also, the current dataset is not large enough to produce meaningful metrics for all classes (particularly if we wish to expand the treatments further). So, next steps are: 
- Expand the dataset to gather all SCOTUS opinions that cited the cited opinions in the current dataset, this should substantially expand the dataset to provide enough examples to produce meaningful metrics. If this produces a dataset that's forbiddingly large, we can then sample from this dataset to reduce its size. 
- Assign subject-matter-annotated labels to the expanded dataset. 
- Enhance prompt with SME guidance.
- Re-run eval once all steps above are done.
- @mlissner might need your help coordinating SME, we might need a number of SMEs to help ease the load, it would be best to have at least 3 subject-matter-annotated labels for each record to reduce bias, but we can start with just splitting the dataset to SMEs without overlap. Also, maybe we can discuss the road forward given AWS project, wouldn't want us to duplicate work.

All work are documented here: https://github.com/freelawproject/ai-experiments/tree/FLP_Citator/experiments_515

**Comment by mlissner** [comment]
> This distinction was not apparent to me previously and I'm not entirely sure why we chose this direction instead of the other one

I don't think we actually considered going the other way. I think this is just because it's less intuitive, but it's not something we ruled out or anything. We can easily identify everything that cites something, but once you have the 1,000 decisions citing something, you'd have to find the paragraphs or sentences within those decisions that actually matter. Maybe that's not so hard, but we don't have an easy way of doing that now. The other thing is that there might be 999 citations of something that are fine, but then SCOTUS comes along and cites it, and suddenly it's bad law. Tricky.

> Expand the dataset to gather all SCOTUS opinions that cited the cited opinions

I think to do this, we'd want to ask the case law team to re-run eyecite on all of SCOTUS. I think this is easy these days.

> @mlissner might need your help coordinating SME,

Sounds good. Let's discuss. Rebecca is onboard, but we'd have to think about others that we could recruit and how to coordinate them.
